src/main/java/com/guysfromusa/carsgame/game_state/GameStateTracker.java

```diff
@@ -19,21 +28,36 @@ public void addNewGame(String gameId){
         this.gameStateMap.put(gameId, new GameState());
     }
 
-    public  void addNewCar(String gameId, String carName){
+    public  void addNewCar(String gameId, String carName) throws NotFoundException {
        GameState gameState = this.gameStateMap.get(gameId);
+       if(gameState == null)
+           throw new NotFoundException(format("Game with id {} dosn't exist", gameId));
        gameState.addNewCar(carName);
     }
 
-    public void addNewMove(String gameId, String carName, Movement.Operation operation){
+    public void addNewMove(String gameId, String carName, Movement.Operation operation) throws NotFoundException {
         GameState gameState = this.gameStateMap.get(gameId);
+        if(gameState == null)
+            throw new NotFoundException(format("Game with id {} dosn't exist", gameId));
         gameState.addNewMovement(carName, operation);
     }
 
     public List<Movement> getCarsMovementHistory(String gameId, String carName){
-       return  this.gameStateMap.get(gameId).getCarsMovement(carName);
+        GameState gameState = this.gameStateMap.get(gameId);
+        return gameState != null ? gameState.getCarsMovement(carName) : emptyList();
     }
 
     public GameState getGameState(String gameId){
         return this.gameStateMap.get(gameId);
     }
+
+    public List<Movement> getNCarsMovementHistory(String gameId, String carName, int numberOfStepBack){
+        GameState gameState = this.gameStateMap.get(gameId);
+        List<Movement> carsMovement = emptyList();
+        if(gameState != null){
+           carsMovement = gameState.getCarsMovement(carName);
+        }
+        return  carsMovement.isEmpty() ? carsMovement :
```

Comment:
we switched to Collection due to lack of efficient concurrent List, please try to implement this with iterator

Reply:
you can switch to COpyOnWriteArrayList